dali/python/nvidia/dali/plugin/tf.py

```diff
@@ -274,12 +287,63 @@ def __init__(self,
 
-        def _check_output_dtypes(self, output_dtypes):
-            """Check whether output_dtypes is instance of tf.DType or tuple of tf.DType"""
-            if isinstance(output_dtypes, tf.DType):
+        def _setup_inputs(self, input_datasets, input_names, input_layouts):
+            """Verify the input specification and assign it to private members in
+            normalized form."""
+            try:
+                nest.assert_same_structure(input_datasets, input_names)
+            except ValueError as e:
+                raise ValueError("`input_datasets` and `input_names` structure do not match.") from e
+            if input_datasets is not None:
+                if not self._check_dtypes(input_datasets, dataset_ops.DatasetV2):
+                    raise TypeError(("`input_datasets` should be provided as a " +
+                        "tf.data.Dataset object or a tuple of Datasets. Got `{}` of type `{}`.") \
+                        .format(input_datasets, type(input_datasets)))
+
+                if not self._check_dtypes(input_names, str):
+                    raise TypeError(("`input_names` should be provided as a str or a tuple" +
+                        " of str. Got `{}` of type `{}`.") \
+                        .format(input_names, type(input_names)))
+
+                # TODO(klecki): we can possibly add a syntactic sugar and allow only one layout to be
+                # specified for all inputs uniformly
+                if input_layouts is not None:
+                    try:
+                        nest.assert_same_structure(input_layouts, input_datasets)
+                    except ValueError as e:
+                        raise ValueError("`input_datasets` and `input_layouts` structure do not match.") from e
+                    if not self._check_dtypes(input_layouts, str):
+                        raise TypeError(("`input_layouts` should be provided as a str " +
+                            "or a tuple of str. Got `{}` of type `{}`.") \
+                            .format(input_layouts, type(input_layouts)))
+                else:
+                    input_layouts = nest.map_structure(lambda _: "", input_datasets)
+            else:
+                # Make them explicitly empty for the internal representations
+                input_datasets = ()
+                input_names = ()
+                input_layouts = ()
+
+            if not isinstance(input_datasets, tuple):
+                input_datasets = (input_datasets, )
+                input_names = (input_names, )
+                input_layouts = (input_layouts, )
+
+            self._input_datasets = input_datasets
+            self._input_names = input_names
+            self._input_layouts = input_layouts
+
+            # TODO(klecki): Validate number of specified inputs against the Pipeline
+
+
+        def _check_dtypes(self, provided_nest, expected_elem_type, allow_single=True):
```

Comment:
1. Why would check_dtype**s** (plural) even allow single value?
2. `provided_nest` - it's not a nest (it's a "nest", it's a tuple) - also, the "provided" part isn't very informative, either
3. It could (and probably should) be a free function:
  a) it's generic (not specific for the enclosing class)
  b) `self` is not used

```suggestion
        def check_dtypes(values, expected_elem_type):
```

Reply:
1. I had an idea to allow only tuples for some of the parameters, but it's not useful. 

For the inputs, this form will be replaced with the dictionary in the next iteration, but we keep it for the output_dtypes.

2. changed to value, I know that this is maybe not most intuitive function name, so it has a docstring.

3. Certainly not a non-private function we want to expose as a part of dali.plugin.tf interface.
It is just for code repetition in this class.